### PR TITLE
Auto generate models

### PR DESCRIPTION
**Pull request by netpok**
This pull request allows generating model helpers automatically after migrations executed.

Basically a newer version of #842, started from the current master branch.

Based on the comment by barryvdh the config option is simplified it is named `post_migrate` and it is either false to disable this feature or sets the parameters for the "ide-helper:model" command. It supports both `"--nowrite"` and the `["--nowrite" => true]` format.

The trigger was updated to use the MigrationsEnded event because Laravel will run "incomplete" commands like `migrate:roll` as `migrate:rollback`.

The other helper file generation was dropped in favor of composer post update commands.

### CODE REVIEW

File: src/Listeners/GenerateModelHelper.php [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+     *
+     * @param  CommandFinished  $event
+     */
+    public function handle(CommandFinished $event)
+    {
+        if (!self::$shouldRun) {
+            return;
+        }
+
+        self::$shouldRun = false;
```
Comment (mfn) [inline src/Listeners/GenerateModelHelper.php:41]: Why is this necessary?

**Review by mfn — COMMENTED** [review]
Generally LGTM, I've a question (inline) though, thanks

File: src/Listeners/GenerateModelHelper.php (hunk 1 above)
Comment (netpok) [reply src/Listeners/GenerateModelHelper.php:41]: To access the current output buffer we need to run the command on the `CommandFinished` event, where we can access the commands output.

To achieve this:
- We set the flag on the `MigrationsEnded` event
- If the flag is set we run the command
- But we need to reset the flag before that otherwise it would also trigger on the end of the generate command resulting in an infinite loop.

The best way to avoid this would be single fired events but as far as I know Laravel does not support them.

Comment (mfn) [reply src/Listeners/GenerateModelHelper.php:41]: Makes sense to me!

Can you please add this as a phpdoc comment to the `$shouldRun` property? I doubt this will be easy to figure out :)

After that I think we're good here.

Comment (netpok) [reply src/Listeners/GenerateModelHelper.php:46]: I added one

**Review by mfn — APPROVED** [review]
LGTM!

**Comment by netpok** [comment]
I came across a project which uses the `spatie/laravel-multitenancy` package, thus running the basic `artisan ide-helper:generate` does not suffice.

Solutions in my mind:
1. Relaxing the constraint by requiring the full command to be entered, e.g: `ide-helper:models -N`, maybe even allowing multiple commands. This solution would mean dropping array parameter compatibility, but I don't see that would be used too much.
2. Allowing to use a custom action, e.g.: allowing to set a callback which defaults to the current implementation `GenerateModelHelper::generateWith(callback);` in AppServiceProvider
3. This is an edge case which the user should handle on it's own, e.g: by overriding the generator listener, which should be possible with the current code by using `$app->register(GenerateModelHelper::class, CustomImplementation::class);`

**Comment by mfn** [comment]
> This solution would mean dropping array parameter compatibility, but I don't see that would be used too much.

At least I can say I agree to that assertion :)

I think I would say, the bottom line for all of this would be improved DX for users of this library. Seems only 1) really fulfills this, once you "have to code" I think it gets worse. Also creates more dependencies on the internal code, which we might rather change one day…

**Comment by netpok** [comment]
I pushed a commit which implements option 1.

File: src/Listeners/GenerateModelHelper.php [hunk 2]
```diff
@@ -45,11 +46,10 @@ public function handle(CommandFinished $event)
...
 
-        $parameters = $this->config->get('ide-helper.post_migrate');
-        $this->artisan->call(
-            is_array($parameters) ? 'ide-helper:models' : 'ide-helper:models ' . $parameters,
-            is_array($parameters) ? $parameters : [],
-            $event->output,
-        );
+        $commands = $this->config->get('ide-helper.post_migrate');
+
+        foreach (Arr::wrap($commands) as $command) {
```
Comment (mfn) [inline src/Listeners/GenerateModelHelper.php:51]: The `Arr::wrap` I think will break when using the current `false`, can you check?

`Arr::wrap` has a special case for `null` coercing to an empty array (which would be fine here), but otherwise it wants to process the defaults as `[false]` when I read this correctly.

Maybe it would be just easier to make `post_migrate` always an array, also in the config, with an empty one but one commented example which could user help to quicker start using it:
```php
'post_migrate' => [
    // 'ide-helper:models --nowrite',
],
```

Comment (mfn) [reply src/Listeners/GenerateModelHelper.php:51]: Further, please also add the default to `$this->config->get()`, not everyone upgrading the library immediately upgrades the config files (though incidentally, in _this_ case the `null` default would just work but it just would seem inconsistent then).

Comment (netpok) [reply src/Listeners/GenerateModelHelper.php:51]: It should not matter since the [event listener is not registered when falsy values are provided](https://github.com/barryvdh/laravel-ide-helper/pull/1163/files#diff-08724b347a720a68357b78ce96d5c46a5fcb0c90bc3784594620a0af65e2bf45R38)

Comment (netpok) [reply src/Listeners/GenerateModelHelper.php]: Simplifying the config makes sense so I did that. Also the package config is always merged with the user config so there shouldn't be an event where the config entry is missing.

File: src/IdeHelperServiceProvider.php [hunk 3]
```diff
@@ -32,6 +35,13 @@ class IdeHelperServiceProvider extends ServiceProvider implements DeferrableProv
      */
     public function boot()
     {
+        if ($this->app['config']->get('ide-helper.post_migrate', false)) {
```
Comment (mfn) [inline src/IdeHelperServiceProvider.php:38]: Thanks for the other change and the clarification! I think for consistency, we also should default to `[]` here now:
```suggestion
        if ($this->app['config']->get('ide-helper.post_migrate', [])) {
```

**Review by mfn — APPROVED** [review]
One last comment, otherwise LGTM!

File: src/Listeners/GenerateModelHelper.php (hunk 2 above)
Comment (mfn) [reply src/Listeners/GenerateModelHelper.php]: > Also the package config is always merged with the user config

That's interesting, we definitely had issues in the past due to missing config values and no defaults in the code 🤷‍♀️

File: src/IdeHelperServiceProvider.php (hunk 3 above)
Comment (netpok) [reply src/IdeHelperServiceProvider.php]: I thought leaving this as false since its a condition but ok, changed it.

File: src/Listeners/GenerateModelHelper.php (hunk 2 above)
Comment (netpok) [reply src/Listeners/GenerateModelHelper.php]: Interesting, this [line should](https://github.com/barryvdh/laravel-ide-helper/blob/master/src/IdeHelperServiceProvider.php#L57) prevent that.

**Comment by mfn** [comment]
@barryvdh IMHO good to merge!

**Comment by netpok** [comment]
@mfn sorry to bother you but could you fix my name in the changelog next time you edit it?
